gridstatus/spp.py

```diff
@@ -193,6 +205,169 @@ def get_load_forecast(self, date, forecast_type="MID_TERM", verbose=False):
+        """
+        Returns solar and wind generation forecast for +7 days in hourly intervals.
+
+        Arguments:
+            date (pd.Timestamp|str): date to get data for. Supports "latest" and "today"
+            verbose (bool): print info
+
+        Returns:
+            pd.DataFrame: forecast as dataframe.
+        """
+        # The MID_TERM forecast is delayed up to 10 minutes.
+        buffer_minutes = 10
+
+        now = self.now() - pd.Timedelta(minutes=buffer_minutes)
+
+        if date in ["latest", "today"]:
+            date = now
+
+        date = utils._handle_date(date, self.default_timezone)
+        date = min(now, date)
+
+        if end:
+            end = utils._handle_date(end, self.default_timezone)
+            # Make sure we don't request files that don't exist yet
+            end = min(now, end)
+
+        df = self._retrieve_solar_and_wind_forecast_mid_term(date, end, verbose)
+        df["Forecast Type"] = "MID_TERM"
+
+        if end:
+            return df[df["Publish Time"].between(date, end)]
+
+        return df
+
+    @support_date_range("HOUR_START")
+    def _retrieve_solar_and_wind_forecast_mid_term(self, date, end=None, verbose=False):
+        """System-wide wind and solar forecast data for +7days by hour."""
+        url = self._mid_term_solar_and_wind_url(date.floor("H"))
+
+        df = pd.read_csv(url)
+
+        df = self._post_process_solar_and_wind_forecast(
+            df,
+            url,
+            end_time_col="GMTIntervalEnd",
+            interval_duration=pd.Timedelta(hours=1),
+        )
+
+        return df
+
+    def _post_process_solar_and_wind_forecast(
+        self,
+        df,
+        url,
+        end_time_col,
+        interval_duration,
+    ):
+        df = self._handle_market_end_to_interval(df, end_time_col, interval_duration)
+
+        # Assume the publish time is in the name of the file. There are different
+        # times on the webpage, but these could be the posting time.
+        df["Publish Time"] = pd.Timestamp(
+            url.split("-")[-1].split(".")[0],
+            tz=self.default_timezone,
+        )
+
+        df.columns = [col.strip() for col in df.columns]
+
+        df = (
+            utils.move_cols_to_front(
+                df,
+                ["Interval Start", "Interval End", "Publish Time"],
+            )
+            .drop(columns=["Time", "Interval"])
+            .sort_values(["Interval Start", "Publish Time"])
+        )
+
+        return df.dropna(subset=["Wind Forecast MW", "Solar Forecast MW"]).reset_index(
+            drop=True,
+        )
+
+    def _short_term_solar_and_wind_url(self, date):
+        hour = date.hour
+        padded_hour = str(hour).zfill(2)
+        padded_hour_plus_one = str((hour + 1) % 24).zfill(2)
+
+        # The first hour in the URL is 1 after the hour in the filename.
+        # Example 2024/01/01/02 has data for 01/01/2024 01:00:00 - 01/01/2024 01:55:00
```

Comment:
could this be because the path in the url is based on the start of the interval, while the file name is based on the end? 

see how i handle this [here in operating reverses](https://github.com/kmax12/gridstatus/blob/c24523bd5eabe7c26fba5cf62a85020d13e76f12/gridstatus/spp.py#L854), by using the start and end to make url. 

if that applies in the situation, might make the code easier to understand if structured that way

Reply:
The URL path is based on the published time of the forecast. For example, if I want to get the forecast published at 0415 on 2024-01-29, then the path is https://portal.spp.org/file-browser-api/download/shortterm-resource-forecast?path=/2024/01/29/05/OP-STRF-202401290415.csv

Note that the first hour in the URL (05) is 1 greater than the publish hour of the forecast (04).

The structure of the operating reserves is different because it doesn't have folders for each hour.

Reply:
i got it now. the name of the folder is the hour ending of the hour that the 5 minute interval is part of. which is taking the hour of the publish time and adding one. 

looks like the logic to handle the wrap around at the end of the day is implemented via the `% 24`. i would just make sure that case is tested since i didnt notice it explicitly when I skimmed tests

Reply:
Added a test to handle the hour 24 case (11 pm). 

https://github.com/kmax12/gridstatus/pull/324/files#diff-1e7d0a4a3e126f70ca1a7d6dd9a41a54a5574da5060cdf19062f47a9dc15c14bR419